Split Codespaces configuration based on development scenarios

eerhardt opened the pull request:
Taking advantage of the new "monorepo" projects feature in https://github.blog/changelog/2022-07-28-prebuilding-codespaces-is-now-supported-for-multi-repository-and-monorepo-projects/.

This allows us to have different Codespaces pre-builds based on what kind of work you are doing. It also minimizes the down-time of pre-builds if one development scenario is broken (for example if the WASM prebuild is broken, libraries developers can still use the libraries pre-build).

The duplication of the `Dockerfile` is unfortunate. But I'm thinking if we can move these docker images to https://github.com/dotnet/dotnet-buildtools-prereqs-docker, we could:
1. save time during the pre-build actions because the image doesn't need to be rebuilt every day.
2. Refactor the Dockerfiles into layers and reuse the common pieces.
cc @MichaelSimons and @mthalman for thoughts on that.

This will require an admin to update the Codespaces in the Settings tab of the repo. I don't have access to do that.

eerhardt committed:
Split libraries and wasm devcontainers

Codespaces allows for monorepo support now. So we can have different pre-builds for different dev environments.

Creating a "libraries" pre-build and a "wasm" pre-build devcontainer.

eerhardt committed:
Move devcontainer files into separate folders

eerhardt committed:
Path up a directory to the Dockerfile

eerhardt committed:
Split the Dockerfile so it can be customized per devcontainer

eerhardt committed:
Update Codespaces docs

mthalman commented on .devcontainer/libraries/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.192.0/containers/dotnet/.devcontainer/base.Dockerfile
+# For details on dotnet specific container, see: https://github.com/microsoft/vscode-dev-containers/tree/main/containers/dotnet
+
+# [Choice] .NET version: 6.0, 5.0, 3.1, 2.1
```
I understand this was generated but probably don't need to have 5.0 and 2.1 referenced.

mthalman commented on .devcontainer/libraries/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+# Set up machine requirements to build the repo and the gh CLI
+RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+    && apt-get -y install --no-install-recommends cmake llvm-10 clang-10 \
+        build-essential python curl git lldb-6.0 liblldb-6.0-dev \
+        libunwind8 libunwind8-dev gettext libicu-dev liblttng-ust-dev \
+        libssl-dev libnuma-dev libkrb5-dev zlib1g-dev ninja-build \
+        gh
```
Normal convention is to have each package name on its own line and listed alphabetically. This allows for more straightforward diffs when changes are made and easier to locate packages.

mthalman commented on .devcontainer/libraries/devcontainer.json:
```diff
@@ -0,0 +1,47 @@
+		"args": {
+			// Update 'VARIANT' to pick a .NET Core version: 3.1, 6.0
+			// Append -bullseye or -focal to pin to an OS version.
+			"VARIANT": "6.0-focal"
```
Any reason that `6.0-jammy` (Ubuntu 22.04) can't be used instead of `6.0-focal` (Ubuntu 20.04)?

samruddhikhandale replied:
For the `dotnet` image, `jammy` is not supported yet. You can see the complete list of available tags [here](https://mcr.microsoft.com/v2/devcontainers/dotnet/tags/list)

radical commented on docs/workflow/Codespaces.md:
```diff
@@ -7,9 +7,20 @@ dotnet/runtime runs a nightly GitHub Action to build the latest code in the repo
+![New codespace button](https://docs.github.com/assets/cb-138303/images/help/codespaces/new-codespace-button.png)
 
-2. Select the Machine type. For dotnet/runtime, it is recommended to select at least a `4-core` machine. You can also verify that a "Prebuild" is ready.
+2. Click the drop-down at the size of the "Create codespace on main" button and select "Configure and create codespace"
```
```suggestion
2. Click the drop-down at the side of the `Create codespace on main` button and select `Configure and create codespace`
```

radical commented on docs/workflow/Codespaces.md:
```diff
@@ -7,9 +7,20 @@ dotnet/runtime runs a nightly GitHub Action to build the latest code in the repo
+
+![Dev container configuration](./codespace-dev-container-configuration.png)
+
+    - For "libraries" work, pick ".devcontainer/libraries/devcontainer.json".
+    - For "WASM" work, pick ".devcontainer/wasm/devcontainer.json".
```
```suggestion
    - For `libraries` work, pick `.devcontainer/libraries/devcontainer.json`.
    - For `WASM` work, pick `.devcontainer/wasm/devcontainer.json`.
```

MichaelSimons commented:
> 2\. Refactor the Dockerfiles into layers and reuse the common pieces.
>     cc @MichaelSimons and @mthalman for thoughts on that.

I think that makes sense. You are installing a lot of common prereq packages that aren't going to change at a high frequency.

eerhardt committed:
Respond to PR feedback

eerhardt commented on docs/workflow/Codespaces.md:
```diff
@@ -20,21 +31,15 @@ dotnet/runtime runs a nightly GitHub Action to build the latest code in the repo
     - The `scripts` folder contains any scripts that are executed during the creation of the codespace. This has the build command that builds the entire repo for prebuilds.
-2. The GitHub Action can be configured by following the instructions at https://docs.github.com/codespaces/prebuilding-your-codespaces/configuring-prebuilds.
+2. Each development scenario folder contains:
+    - `devcontainer.json` file configures the codespace and has VS Code / Environment settings
```
"VS Code / Environment settings" isn't a referenced string. It just means this file has the settings for the VS Code and the environment.

samruddhikhandale commented on .devcontainer/libraries/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+
+# [Choice] .NET version: 6.0, 3.1
+ARG VARIANT="6.0-focal"
+FROM mcr.microsoft.com/vscode/devcontainers/dotnet:0-${VARIANT}
```
```suggestion
FROM mcr.microsoft.com/devcontainers/dotnet:0-${VARIANT}
```

As we're starting to migrate contents of [microsoft/vscode-dev-containers](https://github.com/microsoft/vscode-dev-containers) repo to the [devcontainers org](https://github.com/devcontainers), as part of the work on the [open dev container specification](https://containers.dev).

We'll now be publishing images from [devcontainers/images](https://github.com/devcontainers/images) with path `mcr.microsoft.com/devcontainers/*`. Images with path `mcr.microsoft.com/vscode/devcontainers/*` will still be published from the new repo for backwards compatibility, but it will be nice if you could adapt to it now. 😄 

For more details, you can review the [announcement issue](https://github.com/microsoft/vscode-dev-containers/issues/1589)

eerhardt replied:
I missed this announcement. Thanks for the review. I'll get this changed.

eerhardt committed:
Use the new open devcontainers path.

eerhardt committed:
Use the GH CLI feature instead of installing it ourselves

fanyang-mono reviewed (approved):
WASM related change looks good to me! This is a great new feature. Now I could add the build for Mono desktop scenarios.

eerhardt commented:
I have addressed all feedback and have manually tested this works in my fork. Once I merge, I'm going to need a repo admin to update the settings in order to get the Codespaces prebuilds working again.

I plan on merging tomorrow unless I get more feedback.

eerhardt committed:
Set hostRequirements for Codespaces

Developing in dotnet/runtime with a 2-core / 4GB ram machine doesn't work very well. Add a minimum of 4-core machine to the devcontainer spec.

Fix #75680
